tensorflow/java/src/main/native/server_jni.cc

```diff
@@ -0,0 +1,131 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/java/src/main/native/server_jni.h"
+#include "tensorflow/c/c_api.h"
+#include "tensorflow/java/src/main/native/exception_jni.h"
+#include "tensorflow/java/src/main/native/utils_jni.h"
+
+namespace {
+#ifndef __ANDROID__
+TF_Server* requireHandle(JNIEnv* env, jlong handle) {
+  static_assert(sizeof(jlong) >= sizeof(TF_Server*),
+                "Cannot package C object pointers as a Java long");
+  if (handle == 0) {
+    throwException(env, kIllegalStateException,
+                   "close() has been called on the Server");
+    return nullptr;
+  }
+
+  return reinterpret_cast<TF_Server*>(handle);
+}
+#endif  // __ANDROID__
+
+}  // namespace
+
+JNIEXPORT jlong JNICALL Java_org_tensorflow_Server_allocate(
+    JNIEnv* env, jclass clazz, jbyteArray server_def) {
+#ifdef __ANDROID__
+  throwException(env, kUnsupportedOperationException,
+                 "Server is not supported on Android");
+  return 0;
+#else
+  TF_Status* status = TF_NewStatus();
+
+  jbyte* server_def_ptr = env->GetByteArrayElements(server_def, nullptr);
+
+  TF_Server* server = TF_NewServer(
+      server_def_ptr, static_cast<size_t>(env->GetArrayLength(server_def)),
+      status);
+
+  env->ReleaseByteArrayElements(server_def, server_def_ptr, JNI_ABORT);
+  bool ok = throwExceptionIfNotOK(env, status);
+
+  TF_DeleteStatus(status);
+
+  return ok ? reinterpret_cast<jlong>(server) : 0;
+#endif  // __ANDROID__
+}
+
+JNIEXPORT void JNICALL Java_org_tensorflow_Server_start(JNIEnv* env,
+                                                        jclass clazz,
+                                                        jlong handle) {
+#ifdef __ANDROID__
```

Comment:
Actually, one more thing: Why not put these error messages in `c_api.cc` instead of in the JNI code?

Reply:
Currently this functions are fully excluded from `c_api.cc` and `c_api.h` with their definitions in case we build TensorFlow for Android. Additionally, these exceptions are Java specific (like `NotSupportedOperationException`), so I can't throw them from `c_api.cc` anyway. Of course, I can throw another exception or status from `c_api.cc` and handle it in JNI converting into `NotSupportedOperationException`, but it looks like it will significantly complicate the solution, don't you think so?

Reply:
Okay, @asimshankar, I moved these `#ifdef` constructions into `c_api.cc`, please have a look.